tests/spi/main.c

```diff
@@ -0,0 +1,75 @@
+#include <test/test.h>
+#include <uart/uart.h>
+
+void init_uart_test(void) {
+  init_uart();
+  // set pin 4 to high/output
+  init_cs(4, &DDRB);
+
+  //ASSERT_TRUE(0x10 & var)
+  // isolate 4th position in DDRB
+  uint8_t pin = DDRB&(1<<4);
+  //shift value to 0th position
+  pin >>= 4;
+  //check if it's high ie equals 1
+  ASSERT_EQ(pin,1);
+// or ASSERT_TRUE(pin); means it can be anything but 0 so use ASSERT_EQ because wants it to be just one
+}
+
+void set_cs_low_test(void) {
+  //check set cs low/for input (so slave is listening)
+  set_cs_low(4,&PORTB);
+  pin = PORTB&(1<<4);
+  pin >>= 4;
+  ASSERT_EQ(pin,0);
+}
+
+void set_cs_high_test(void) {
+  //check set cs high/for output (so not listening)
+  set_cs_high(4,&PORTB);
+  pin = PORTB&(1<<4);
+  pin >>= 4;
+  ASSERT_EQ(pin,1);
+}
+
+void init_spi_test_1(void) {
+  //check init_spi
+  init_spi();
+  //check MOSI and CLK are output (ie high)
+  uint8_t pins = DDRB & (_BV(MOSI) + _BV(CLK));
+  ASSERT_EQ(pins,(_BV(MOSI) + _BV(CLK)));
+}
+void init_spi_test_2(void) {
+  init_spi();
+  //check SS is output (ie high)
+  uint8_t SS_pin = DDRD & _BV(SS);
+  ASSERT_EQ(SS_pin,_BV(SS));
+}
+void init_spi_test_3(void) {
+  init_spi();
+  //check
+  uint8_t pins = SPCR & (_BV(SPE) + _BV(MSTR) + _BV(SPR1));
```

Comment:
Instead of using `_BV(..) + _BV(..)` you should use `_BV(..) | _BV(..)`; in some cases, these can do the same thing, but in most cases, you want to avoid the carryover that comes with +. For example, 001 + 001 = 010, yet 001 | 001 = 001. In the first case, the "carry the one" behaviour is not desired.

Reply:
Otherwise, this test looks great. @jackberezny this is the kind of format I'm hoping to use, of course, with CAN this will become a little more complicated, but the idea is to have small-ish self-contained tests.